src/Sign.SignatureProviders.TrustedSigning/TrustedSigningService.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE.txt file in the project root for more information.
+
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Azure;
+using Azure.CodeSigning;
+using Azure.CodeSigning.Models;
+using Azure.Core;
+using Microsoft.Extensions.Logging;
+using Sign.Core;
+
+namespace Sign.SignatureProviders.TrustedSigning
+{
+    internal sealed class TrustedSigningService : ISignatureAlgorithmProvider, ICertificateProvider, IDisposable
+    {
+        private static readonly SignRequest _emptyRequest = new(SignatureAlgorithm.RS256, new byte[32]);
+
+        private readonly CertificateProfileClient _client;
+        private readonly string _accountName;
+        private readonly string _certificateProfileName;
+        private readonly ILogger<TrustedSigningService> _logger;
+        private readonly SemaphoreSlim _mutex = new(1);
+        private X509Certificate2? _certificate;
+
+        public TrustedSigningService(
+            TokenCredential tokenCredential,
+            Uri endpointUrl,
+            string accountName,
+            string certificateProfileName,
+            ILogger<TrustedSigningService> logger)
+        {
+            ArgumentNullException.ThrowIfNull(tokenCredential, nameof(tokenCredential));
+            ArgumentNullException.ThrowIfNull(endpointUrl, nameof(endpointUrl));
+            ArgumentException.ThrowIfNullOrEmpty(accountName, nameof(accountName));
+            ArgumentException.ThrowIfNullOrEmpty(certificateProfileName, nameof(certificateProfileName));
+            ArgumentNullException.ThrowIfNull(logger, nameof(logger));
+
+            _accountName = accountName;
+            _certificateProfileName = certificateProfileName;
+            _logger = logger;
+
+            _client = new CertificateProfileClient(tokenCredential, endpointUrl);
+        }
+
+        public void Dispose()
+        {
+            _mutex.Dispose();
+            _certificate?.Dispose();
+            GC.SuppressFinalize(this);
+        }
+
+        public async Task<X509Certificate2> GetCertificateAsync(CancellationToken cancellationToken)
+        {
+            if (_certificate is not null)
+            {
+                return new X509Certificate2(_certificate);
+            }
+
+            await _mutex.WaitAsync(cancellationToken);
+            try
+            {
+                if (_certificate is null)
+                {
+                    Stopwatch stopwatch = Stopwatch.StartNew();
+
+                    _logger.LogTrace(Resources.FetchingCertificate);
```

Comment:
`Resources.FetchingService` returns:

`Fetching certificate from Azure Key Vault.`

We could either flavor this message for each certificate provider or make the message more generic (e.g.:  `Fetching certificate.`).

The real problem is that Sign.SignatureProviders.TrustedSigning uses Sign.Core.Resources.  While this is temporarily OK, Sign.SignatureProviders.TrustedSigning should not directly reference Sign.Core.Resources; each should have its own resource strings.

Because we lack an extensibility model it's passable to reference Sign.Core.Resources for now.  Making the message more generic is OK now, but the better solution would be to create a separate string resources.

I notice that [`CertificateStoreService`](https://github.com/dotnet/sign/blob/db11f797c7c1ba2d133a63474a0d458f9dc178df/src/Sign.SignatureProviders.CertificateStore/CertificateStoreService.cs#L115) has the same problem.

Reply:
That's why I thought it was safe to copy this resource. I haven't looked at the content of those message as you and Joel noticed. I will add a resource file for the Trusted Signing. For the other two SignatureProviders I will make those changes in the other PR that changes the DI of the service.